File: leetcode/medium/98-validate-bst/slow-rec.py
"""
  @author Vortexx2
  Problem 98 - Validate BST

  Runtime - 124 ms O(n^2)
  Memory Usage - 17 MB O(n^2)
"""
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def isValidBST(self, root: Optional[TreeNode]) -> bool:
    if not root:
      return True
    self.val = root.val

    if (not self.isLesser(root.left) or not self.isGreater(root.right)):
      return False

    return self.isValidBST(root.left) and self.isValidBST(root.right)

  def isLesser(self, root: TreeNode):
    if not root:
      return True

    return root.val < self.val and self.isLesser(root.left) and self.isLesser(root.right)

  def isGreater(self, root: TreeNode):
    if not root:
      return True

    return root.val > self.val and self.isGreater(root.left) and self.isGreater(root.right)

File: leetcode/medium/98-validate-bst/slow-rec.cpp
/**
 * @file slow-rec.cpp
 * @author Vortexx2
 * @brief Problem 98 - Validate BST
 * @date 29-08-2021
 * 
 * Runtime - 8 ms O(n^2)
 * Memory Usage - 21.6 MB O(n^2) [recursive stack]
 */

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  bool isValidBST(TreeNode *root) {
    if (!root) return true;
    if (!isLesser(root->left, root->val) || !isGreater(root->right, root->val)) return false;

    return isValidBST(root->left) && isValidBST(root->right);
  }

  bool isLesser(TreeNode *root, int &value) {
    if (!root) return true;

    return (root->val < value) && isLesser(root->left, value) &&
           isLesser(root->right, value);
  }

  bool isGreater(TreeNode *root, int &value) {
    if (!root) return true;

    return (root->val > value) && isGreater(root->left, value) &&
           isGreater(root->right, value);
  }
};